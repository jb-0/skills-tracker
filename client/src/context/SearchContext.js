/**
  * This context is used to pass state to the numerous components and events (such as save) that
  * make up a search. This also means a user can leave the search screen to view their profile
  * without losing search information.
  */
import React, { createContext, useEffect, useState  } from 'react';
import getPermittedTerms from '../services/getPermittedTerms';

export const SearchContext = createContext();

export const SearchProvider = (props) => {
  // Store a number of variables in state, firstly the input from the search text box, second the
  // terms that have been added so far and finally the suggested search terms
  const [search, setSearch] = useState({
    searchInputText: '',
    searchTerms: [],
    suggestedTerms: [],
    submittedSearchTerms: [],
    location: 'London',
    submittedLocation: '',
    permittedTerms: {locations: [], skills: []},
    handleDropDownSelectUpdates,
    handleSearchButtonClick
  });

  useEffect(() => {
    async function setPermittedTerms() {
      const permittedTerms = await getPermittedTerms();

      setSearch((previousValues) => {
        return { ...previousValues, permittedTerms };
      });
    }
    
    setPermittedTerms()
  }, []);

  function handleDropDownSelectUpdates(event) {
    const target = event.target.name;
    const value = event.target.value;

    setSearch((previousValues) => {
      return { ...previousValues, [target]: value };
    });
  }

  // When a user hits search ensure the location and searchTerms are current
  function handleSearchButtonClick() {
    setSearch((previousValues) => {
      return {
        ...previousValues,
        submittedSearchTerms: [...previousValues.searchTerms],
        submittedLocation: previousValues.location,
      };
    });
  }
  
  return <SearchContext.Provider value={ [search, setSearch] }>{props.children}</SearchContext.Provider>;
};
