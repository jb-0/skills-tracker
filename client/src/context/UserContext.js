/**
 * The user context is used to carry state that determines whether or not a user details such as
 * whether or not a user is currently authenticated
 */
import React, { createContext, useState } from 'react';
import isAuthenticated from '../services/auth';

export const UserContext = createContext();

export const UserProvider = (props) => {
  // Manage state for two items, firstly checkingAuthentication which is indicative of whether
  // or not LOADING should be displayed, once this is false then whether or not a user is auth'd
  // is known and the user hits the router. This also supplies the function that allows state
  // to be re-confirmed
  const [userState, setUserState] = useState({
    checkingAuthentication: true,
    authenticated: false,
    isAuthenticated,
  });

  return (
    <UserContext.Provider value={[userState, setUserState]}>
      {props.children}
    </UserContext.Provider>
  );
};
