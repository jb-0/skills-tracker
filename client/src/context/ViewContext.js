/**
  * Being able to get the screen dimensions in any component is useful for conditionally rendering
  * based on device type as well as changing class names for styling purpose. This context is used
  * to provide this dimension information as well as classify whether the width is mobile or desktop
  */
import React, { createContext, useState, useLayoutEffect } from 'react';

// Determine device type based on width
function sizeName(width) {
  if (width <= 480) return 'Mobile'
  if (width <= 770) return 'Tablet'
  return 'Desktop'
}

// Set initial window sized on creating context
export const ViewContext = createContext({width: window.innerWidth, 
  height: window.innerHeight, 
  device: sizeName(window.innerWidth)});

export const ViewProvider = (props) => {
  const [size, setSize] = useState();

  useLayoutEffect(() => {
    function updateSize() {
      setSize({ 
        width: window.innerWidth, 
        height: window.innerHeight, 
        device: sizeName(window.innerWidth)
      });
    }

    window.addEventListener('resize', updateSize);
    updateSize();

    return () => window.removeEventListener('resize', updateSize);
  }, []);
  return <ViewContext.Provider value={ size }>{props.children}</ViewContext.Provider>;
};
