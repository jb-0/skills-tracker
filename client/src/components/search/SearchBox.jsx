import React, { useContext, useRef } from 'react';
import './SearchSuggestion';
import { SearchContext } from '../../context/SearchContext';
import { SearchInput } from './SearchBox.Styles'

function Search(props) {
  const [search, setSearch]  = useContext(SearchContext);
  const searchBoxRef = useRef();
  // Handle changes to the text box and provide suggested terms
  function handleTextBoxUpdates(event) {
    const target = event.target.name;
    const value = event.target.value;

    let suggested;
    if (value.length > 0) {
      suggested = search.permittedTerms.skills.filter(
        (word) =>
          word.substring(0, value.length).toLowerCase() === value.toLowerCase()
      );
    } else {
      suggested = [];
    }

    setSearch((previousValues) => {
      return { ...previousValues, [target]: value, suggestedTerms: suggested };
    });
  }

  // Scroll down on focus to ensure user see's suggestions and selected items (especially on mobile)
  function handleOnFocus(){
    searchBoxRef.current.scrollIntoView({ behavior: 'smooth' });
  }

  return (
      <SearchInput
        ref={searchBoxRef}
        onChange={handleTextBoxUpdates}
        onFocus={handleOnFocus}
        type="text"
        id="searchInputText"
        name="searchInputText"
        value={search.searchInputText}
        placeholder="Start typing a skill.."
        autoComplete="off"
      ></SearchInput>
  );
}

export default Search;
