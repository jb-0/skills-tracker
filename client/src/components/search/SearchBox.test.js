import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import userEvent from '@testing-library/user-event';

import SearchBox from './SearchBox';
import { SearchProvider } from '../../context/SearchContext';

describe('SearchBox component', () => {
  let container = null;
  beforeEach(async () => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);

    await act(async () => {
      const promise = Promise.resolve()
      render(
        <SearchProvider>
          <SearchBox />
        </SearchProvider>,
        container
        );
        await act(() => promise);
    });
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('search box is successfully rendered', () => {
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  it('the bound value matches user typed input', async () => {
    // Imitate user typing each letter of Jav and assert it is displayed in textbox
    await userEvent.type(screen.getByRole('textbox'), 'Jav');
    expect(screen.getByDisplayValue('Jav')).toBeInTheDocument();
  });

  it('renders correctly when passed arrays containing suggested terms and search terms', async () => {
    const {act} = renderer;
    const promise = Promise.resolve()

    const tree = renderer
    .create(
      <SearchProvider>
          <SearchBox />
        </SearchProvider>
    );

    await act(() => promise);

    expect(tree.toJSON()).toMatchSnapshot()
  });
});
