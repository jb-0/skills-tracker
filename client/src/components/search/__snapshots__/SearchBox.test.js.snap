// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox component renders correctly when passed arrays containing suggested terms and search terms 1`] = `
<input
  autoComplete="off"
  className="sc-gsTCUz gmbZjy"
  id="searchInputText"
  name="searchInputText"
  onChange={[Function]}
  onFocus={[Function]}
  placeholder="Start typing a skill.."
  type="text"
  value=""
/>
`;
